kybra/compiler/kybra_generate/src/canister_method/post_upgrade_method/build.rs

```diff
@@ -13,64 +16,121 @@ use crate::{
     },
     method_utils::params::InternalOrExternal,
     py_ast::PyAst,
+    source_map::SourceMapped,
     Error,
 };
 
 impl PyAst {
-    pub fn build_post_upgrade_method(
-        &self,
-    ) -> Result<(PostUpgradeMethod, TokenStream), Vec<Error>> {
+    pub fn build_post_upgrade_method(&self) -> Result<PostUpgradeMethod, Vec<Error>> {
+        let init_function_defs = self.get_canister_stmt_of_type(CanisterMethodType::Init);
         let post_upgrade_function_defs =
             self.get_canister_stmt_of_type(CanisterMethodType::PostUpgrade);
+        let (init_params, (post_upgrade_params, guard_function_name), post_upgrade_method_body) = (
+            extract_init_function(&init_function_defs),
+            extract_post_upgrade_function(&post_upgrade_function_defs),
+            generate_post_upgrade_method_body(
+                &self.entry_module_name,
+                init_function_defs.get(0),
+                post_upgrade_function_defs.get(0),
+            ),
+        )
+            .collect_results()?;
+        let post_upgrade_method_params =
+            select_post_upgrade_method_params(init_params, post_upgrade_params);
+
+        Ok(PostUpgradeMethod {
+            params: post_upgrade_method_params,
+            body: post_upgrade_method_body,
+            guard_function_name,
+        })
+    }
+}
+
+fn extract_init_function(
```

Comment:
If it were me I would go with the term `build` not `extract`. Consider if the names here are consistent and what you want.